org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/dfs/DfsBlockCaffeineCache.java

```diff
@@ -0,0 +1,165 @@
+package org.eclipse.jgit.internal.storage.dfs;
+
+import com.github.benmanes.caffeine.cache.*;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+public final class DfsBlockCaffeineCache extends DfsBlockCache {
+
+    /** Pack files smaller than this size can be copied through the cache. */
+    private final long maxStreamThroughCache;
+
+    /**
+     * Suggested block size to read from pack files in.
+     * <p>
+     * If a pack file does not have a native block size, this size will be used.
+     * <p>
+     * If a pack file has a native size, a whole multiple of the native size
+     * will be used until it matches this size.
+     * <p>
+     * The value for blockSize must be a power of 2.
+     */
+    private final int blockSize;
+
+    /** Cache of pack files, indexed by description. */
+    private final Cache<DfsPackDescription, DfsPackFile> packFileCache;
+
+    /** Cache of Dfs blocks. */
+    private final Cache<DfsPackKeyWithPosition, Ref> dfsBlockCache;
+
+
+    private DfsBlockCaffeineCache(DfsBlockCaffeineCacheConfig cacheConfig) {
+        // this is an estimation since it's hard to evaluation the actual size of the indices in packFiles
+        long cacheEntrySize = (cacheConfig.getCacheMaximumSize() / cacheConfig.getBlockSize()) / 2;
+
+        maxStreamThroughCache = (long) (cacheConfig.getCacheMaximumSize() * cacheConfig.getStreamRatio());
+
+        blockSize = cacheConfig.getBlockSize();
+
+        packFileCache = Caffeine.newBuilder()
+                .removalListener((DfsPackDescription description, DfsPackFile packFile, RemovalCause cause) ->
+                        packFile.close())
+                .maximumSize(cacheEntrySize)
+                .expireAfterAccess(cacheConfig.getPackFileExpireSeconds(), TimeUnit.SECONDS)
+                .recordStats()
```

Comment:
Think we want to use tritium to track it but it's not open sourced yet? Will add a TODO to add it when we move it internally.